### PR TITLE
Fix app crash when updating UI after sync

### LINKED ISSUE #2953: System.ArgumentOutOfRangeException during initial sync
## 💥 Crash Report
It looks like a bad update on the main log during the initial sync

### 👥 Impact
It was introduced on 1.9. It seems I'm the only affected until now.

### 🔗 Context
It seems to have been introduced on 1.9. I updated from TestFlight and got the error.

### 🛰 Link
https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/groups/d05e64021859943a19911314a43057e752500aa9?appVersion=1.9

### 📝  Exception & Stack Trace
`Exception type and message`
```cs
SectionedListTableViewSource`2[TModel,TCell].RowsInSection (UIKit.UITableView tableview, System.nint section)
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. Parameter name: index

1 ThrowHelper.ThrowArgumentOutOfRange_IndexException ()
2 List`1[T].get_Item (System.Int32 index)
3 ObservableGroupedOrderedCollection`1[TItem].get_Item (System.Int32 index)
4 SectionedListTableViewSource`2[TModel,TCell].RowsInSection (UIKit.UITableView tableview, System.nint section)
5 (wrapper managed-to-native) ObjCRuntime.Messaging.objc_msgSendSuper(intptr,intptr)
6 UITableView.EndUpdates ()
7 ReactiveTableViewBinder`2[TModel,TCell].handleCollectionChanges (System.Collections.Generic.IEnumerable`1[T] changes)
8 AnonymousSafeObserver`1[T].OnNext (T value)
9 ObserveOn`1+Context+_[TSource].OnNextPosted (System.Object value)
10 UIKitSynchronizationContext+<>c__DisplayClass1_0.<Post>b__0 ()
11 NSAsyncActionDispatcher.Apply ()
12 (wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
13 UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
14 UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
15 Application.Main (System.String[] args)
```

### Steps to reproduce:
1) log into the app (1.9)
2) sync the app
3) close the app
4) using the webapp, add at least 2 TEs to some day
5) open the app (sync will start)
6) :boom:

### PR DESCRIPTION
**Pull request by simonrozsival**
## What's this?
This PR changes the way we manipulate with the UI in a way which does not cause `NSInternalInconsistencyException`s (wrong number of sections/rows after the update).

### Relationships
Closes #2953 

We already dealt with this bug during Spring and it was finally solved in #1848, but it was re-introduced after the Rx refactoring of the main view in #2743.

## Why do we want this?
We want the app to work. Not to crash.

## How is it done?
OK. The most important part of the code is in `SectionedListTableViewSource`. The thing is that we cannot change the data structure which is displayed in the UI in any other place than between `tableView.BeginUpdates()` and `tableView.EndUpdates()` otherwise Apple will crash our app. This means that we have to have two collections - first one is the _real_ data and the other one is the _displayed_ data.

When we add/remove/update anything in the collection which is of type `ObservableGroupedOrderedCollection`, we observe changes to the data - inserts, removals, moving row from one place to another, regular updates. We can process the changes one by one and modify the displayed data and the table view at the same time in the same block ("transaction").

There are some tricky bits because our data structure is a list which has indexes. If we have for example a list of 5 sections and we want to insert new row into section with index 3, we have to distinguish between the case when we want to add the item to section 3 and when we want to insert new section at index 3 and increase the number of sections to 6. We also must not forget to remove empty sections. This is so much fun.

I need to add more unit tests for the observable collection and the view source

~We can decide to create tests in a separate PR if we decide that this works well based on functional testing so we can release 1.9 ASAP (before I go on vacation next week).~ There is an issue for creating unit tests: #2979 

### Why not another way?
I would like to make it more elegant. But at least it works now. Suggestions are welcome.

### Side effects
This might break Android (although it builds correctly) and and specifically will have to be reflected in #2971.

## Review hints
Open the web app and the mobile app side by side. Make various changes in the web app and pull the data to the mobile app. Try moving time entries between different days, deleting them, adding new TEs to days where there are already some data and where there is so far no data.

Test that creating, deleting, editing (mainly changing the start date and star time) within the app works well and updates the user interface correctly.

Test that Android is not broken because of these changes.

## :squid: Permissions
Squash whenever with the approval of the owner of the target branch.

### CODE REVIEW

**Comment by khoi** [comment]
I've been playing around with the iOS and haven't had any problems so far. 

Maybe for a problem like this, we can use something similar to `DiffUtil` in Android. I think @daividssilverio mentioned it the other day.

It compares the two datasets and algorithmically produces the change steps (add, move, insert, delete)  that need to be made to the original one for it to be the same.

Some references that I found:
- https://github.com/skanmera/NetDiff
- https://developer.android.com/reference/android/support/v7/util/DiffUtil 
- https://github.com/Instagram/IGListKit
- https://gist.github.com/ndarville/3166060

**Comment by simonrozsival** [comment]
Generating the diff is not a problem as we observe every change to the collection add/delete/update and process each of them individually. The problem is how to take each of these changes and apply it to the UI. For some reason it is so complicated. Maybe our abstractions are too complex.

**Comment by sofokli1** [comment]
Not sure if the following are related to this PR but I think it is a good place to mention and see if these need separate issues
- I found that the web interface wont show time entries that start and finish in the future but the app shows them
- When changing project colors in web and syncing I do not see the colors changing in the app (full restart fixes this)

Making changes to the order of things (delete, change date) seems to work ok. The only issues I see is the order of entries that both start before now and both finish after now are not in the correct order in the app (restart doesn't help).

**Comment by dshr** [comment]
Works with no problems on Android when using the changes from #2950 👍

File: Toggl.Daneel/Extensions/Binders/ReactiveTableViewBinder.cs [hunk 1]
```diff
@@ -100,5 +102,70 @@ public void Dispose()
...
+        {
+            var oldSectionWasRemoved = remove(change.OldIndex);
+            tableView.InsertRows(change.Index.ToIndexPaths(), UITableViewRowAnimation.Automatic);
+
+            if (oldSectionWasRemoved)
+            {
+                // The index of the section which needs refreshing must be the index before the changes happen.
+                // The index of the target section might change if the old section was in front of the target one
+                // and the old one was removed. The original index of the target section would then be a one more
+                // than the index after hte operation (which is the value of `change.Index.Section`).
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Daneel/Extensions/Binders/ReactiveTableViewBinder.cs]: I forgot to push the 💄 PR :-/ Hopefully somebody will fix this some day.

[the remark this answers is not in the document]
Comment (sofokli1) [reply Toggl.Daneel/Extensions/Binders/ReactiveTableViewBinder.cs]: we will survive

**Pull request merged**